### ISSUE TITLE
Add dual-stick shooter example

### ISSUE THREAD

**Issue by SirGnip** [pvcraven/arcade#301]
I'm learning arcade and thought I'd push up one of my simple learning projects for inclusion as an official example.  It tries to follow the "feel" of other examples.  Also uses a snippet or two from other examples (ex: follow_sprite).

**Comment by pvcraven** [comment]
This seems to run ok by keyboard. I do note there's no 'game over' that is displayed. 

On my gamepad, I can only shoot one direction and the ship does not face the same direction it is moving?

**Comment by SirGnip** [comment]
Thanks for the feedback.

By "shoot one direction", do you mean you can only shoot "left", or you can only shoot "horizontal"? 

I'm assuming it is a joystick axis mapping issue.  Windows USB gamepads are notoriously non-standard.  I'm working on adding some debug logging and will push up another PR for you to try again and post the log output back to me.  Most gamepad games should really have a joystick setup routine at the beginning ("press the stick for movement to the left/right/up/down", "press the stick for shooting to the left/right/up/down") to make joystick configuration dynamic.  This actually would be another good example script for arcade.  Anyone who wants to make a game that supports different models of joystick will have to do this.  I'll see if I can make a simple example of this (dynamically mapping buttons and axes).

**Comment by SirGnip** [comment]
Pushed up https://github.com/pvcraven/arcade/pull/308 which fixes to ship rotation (copy/paste error as I copied this code from a slightly more complicated example with my own images, which were rotated differently).  And, it adds a simple game over message.

Thinking about the non-shooting bug.  Does your gamepad have an "analog" button?  Make sure that is pressed.  This will usually turn on a light.  What gamepad are you using?  

Also, if the "analog" button doesn't help, I put some joystick debug logging in the PR (dumping out all axes, buttons, hats, etc).  Can you run the game from the PR, move and shoot with the left and right stick for a few moments and post the output to this thread?  I want to see what controls are receiving signals on your gamepad.

**Comment by SirGnip** [comment]
@pvcraven , have you been able to test the newest [dual_stick_shooter.py](https://github.com/pvcraven/arcade/blob/master/arcade/examples/dual_stick_shooter.py) after merging my recent changes (3a7f152) in?  I'm curious to take a look at the stdout that is generated.

Also, if anyone else can test this and send feedback on what works or doesn't work (and post stdout to this thread), that would be helpful.

**Comment by pvcraven** [comment]
Not yet. I'm under contract for writing a book, and it has sucked up a lot of my time.

I really want to fix a bug on the Macs, and then maybe I can release 2.0.

**Comment by SirGnip** [comment]
Understood.  Best of luck!  No rush on my account.

**Comment by pvcraven** [comment]
Belatedly getting back to this. I can move the ship just fine, but I only can get an angle of 180 when I shoot. Is it supposed to use the other analog controller to shoot? Or the d-pad?

**Comment by SirGnip** [comment]
I just tested this with my USB gamepad running from the release_2_2_5 tag, and "it worked fine for me."

If no joystick is detected when the example starts, it should use the keyboard (WASD to move, arrow keys to shoot).

If a joystick is detected when the example starts, it uses the left analog stick to move and the right analog stick to shoot.

Now, joysticks/gamepads are notoriously non-standard.  They all label their axes in different ways.  Looking at the `Joystick` object coming back from `pyglet.input.get_joysticks()`, it has `.x`, `.y`, `.z`, `.rx`, `.ry` and `.rz` members.  I was hoping that pyglet did some work to try to map all the different joysticks into these standard x/y/z/rx/ry/rz.  But, no such luck if you are having issues...

The example includes some debugging output to help with this.  Can you run the example, move and shoot with your joystick for a couple seconds, and then post the console output to this thread?

(The solution to make this work on "all" joysticks is to make configuration dynamic by prompting the user "move your left stick side-to-side" and the game detects which axis is moving.  Then, prompt "move your left stick up-and-down", etc. It is a pain.  But other than having a comprehensive database of vendor axes mapping, I know of no other way.)

**Comment by pvcraven** [comment]
Left stick moves the player. Right stick does nothing. Output:

```
   140  0.00 -0.00  0.00  0.95  0.55  0.00  0  0 False False False False False False False False False False
   146  0.00 -0.00  0.00  0.88 -0.63  0.00  0  0 False False False False False False False False False False
   152  0.00 -0.00  0.00  0.13 -1.00  0.00  0  0 False False False False False False False False False False
   158  0.00 -0.00  0.00 -1.00  0.02  0.00  0  0 False False False False False False False False False False
   164  0.00 -0.00  0.00  0.27  1.00  0.00  0  0 False False False False False False False False False False
   170  0.00 -0.00  0.00  0.99  0.19  0.00  0  0 False False False False False False False False False False
   176  0.00 -0.00  0.00  0.94 -0.62  0.00  0  0 False False False False False False False False False False
   182  0.00 -0.00  0.00  0.74 -0.82  0.00  0  0 False False False False False False False False False False
   188  0.00 -0.00  0.00 -0.03  0.01  0.00  0  0 False False False False False False False False False False
   194  0.00 -0.00  0.00 -0.03  0.01  0.00  0  0 False False False False False False False False False False
   200  0.00 -0.00  0.00 -0.03  0.01  0.00  0  0 False False False False False False False False False False
   206  0.00 -0.00  0.00 -0.03  0.01  0.00  0  0 False False False False False False False False False False
   212  0.00 -0.00  0.00 -0.03  0.01  0.00  0  0 False False False False False False False False False False
```

**Comment by SirGnip** [comment]
Could you paste in the whole output?  The initial stuff is useful as well.

**Comment by pvcraven** [comment]
First I moved the do-nothing right stick, then I moved the left stick that works for movement.
```
C:\Users\craven\Desktop\WebServer\arcade\venv\Scripts\python.exe C:/Users/craven/Desktop/WebServer/arcade/arcade/examples/dual_stick_shooter.py
========== <pyglet.input.base.Joystick object at 0x153767F0>
x       0
y       0
z       0
rx      0
ry      0
rz      0
hat_x   0
hat_y   0
buttons [False, False, False, False, False, False, False, False, False, False]
========== Extra joy
button_controls                = [Button(raw_name=Button 0), Button(raw_name=Button 1), Button(raw_name=Button 2), Button(raw_name=Button 3), Button(raw_name=Button 4), Button(raw_name=Button 5), Button(raw_name=Button 6), Button(raw_name=Button 7), Button(raw_name=Button 8), Button(raw_name=Button 9)] (list)
buttons                        = [False, False, False, False, False, False, False, False, False, False] (list)
device                         = DirectInputDevice(name=Controller (Gamepad F310)) (DirectInputDevice)
hat_x                          = 0 (int)
hat_x_control                  = AbsoluteAxis(name=hat, raw_name=Hat Switch) (AbsoluteAxis)
hat_y                          = 0 (int)
hat_y_control                  = AbsoluteAxis(name=hat, raw_name=Hat Switch) (AbsoluteAxis)
rx                             = 0 (int)
rx_control                     = AbsoluteAxis(name=rx, raw_name=X Rotation) (AbsoluteAxis)
ry                             = 0 (int)
ry_control                     = AbsoluteAxis(name=ry, raw_name=Y Rotation) (AbsoluteAxis)
rz                             = 0 (int)
rz_control                     = None (NoneType)
x                              = 0 (int)
x_control                      = AbsoluteAxis(name=x, raw_name=X Axis) (AbsoluteAxis)
y                              = 0 (int)
y_control                      = AbsoluteAxis(name=y, raw_name=Y Axis) (AbsoluteAxis)
z                              = 0 (int)
z_control                      = AbsoluteAxis(name=z, raw_name=Z Axis) (AbsoluteAxis)
========== Extra joy.device
_device                        = <pyglet.libs.win32.dinput.IDirectInputDevice8 object at 0x0D5F8580> (IDirectInputDevice8)
_subtype                       = 512 (int)
_type                          = 21 (int)
controls                       = [AbsoluteAxis(name=y, raw_name=Y Axis), AbsoluteAxis(name=x, raw_name=X Axis), AbsoluteAxis(name=ry, raw_name=Y Rotation), AbsoluteAxis(name=rx, raw_name=X Rotation), AbsoluteAxis(name=z, raw_name=Z Axis), Button(raw_name=Button 0), Button(raw_name=Button 1), Button(raw_name=Button 2), Button(raw_name=Button 3), Button(raw_name=Button 4), Button(raw_name=Button 5), Button(raw_name=Button 6), Button(raw_name=Button 7), Button(raw_name=Button 8), Button(raw_name=Button 9), AbsoluteAxis(name=hat, raw_name=Hat Switch)] (list)
display                        = None (NoneType)
is_open                        = False (bool)
manufacturer                   = None (NoneType)
name                           = Controller (Gamepad F310) (str)
========== pprint joy
<pyglet.input.base.Joystick object at 0x153767F0>
========== pprint joy.device
DirectInputDevice(name=Controller (Gamepad F310))
Using joystick controls: DirectInputDevice(name=Controller (Gamepad F310))
     3  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
     8  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    14  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    20  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    26  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    32  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    38  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    44  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    50  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    56  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    62  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    68  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    74  0.00 -0.00  0.00  0.10  0.01  0.00  0  0 False False False False False False False False False False
    80  0.00 -0.00  0.00  0.99  0.65  0.00  0  0 False False False False False False False False False False
    86  0.00 -0.00  0.00  0.09  1.00  0.00  0  0 False False False False False False False False False False
    92  0.00 -0.00  0.00 -0.55  1.00  0.00  0  0 False False False False False False False False False False
    98  0.00 -0.00  0.00 -1.00  0.40  0.00  0  0 False False False False False False False False False False
   104  0.00 -0.00  0.00 -0.89 -0.55  0.00  0  0 False False False False False False False False False False
   110  0.00 -0.00  0.00 -0.07 -1.00  0.00  0  0 False False False False False False False False False False
   116  0.00 -0.00  0.00  0.77 -0.83  0.00  0  0 False False False False False False False False False False
   122  0.00 -0.00  0.00  1.00  0.06  0.00  0  0 False False False False False False False False False False
   128  0.00 -0.00  0.00  1.00  0.36  0.00  0  0 False False False False False False False False False False
   134  0.00 -0.00  0.00 -0.46  0.11  0.00  0  0 False False False False False False False False False False
   140  0.00 -0.00  0.00 -1.00 -0.00  0.00  0  0 False False False False False False False False False False
   146 -0.81 -0.00  0.00 -0.52 -0.22  0.00  0  0 False False False False False False False False False False
   151 -1.00 -0.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   157 -1.00 -0.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   163 -1.00  0.52  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   169 -0.91  0.68  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   175 -0.91  0.69  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   181 -0.37  0.96  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   187  0.05  1.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   193  0.78  0.87  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   199  0.99  0.34  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   205  0.93 -0.40  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   211  0.00 -1.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   217 -0.36 -1.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   223 -0.95 -0.69  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   229 -0.92 -0.64  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   235 -0.77 -0.75  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   241 -1.00 -0.44  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   247 -0.76 -0.65  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   253  0.63 -1.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   259  0.64 -1.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   265  0.64 -1.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   271  0.64 -1.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   277  0.38 -1.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   283  0.24 -1.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   289  0.23 -1.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   295  0.23 -1.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   301  0.32 -1.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   307  0.52 -1.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   313  0.00 -0.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   319  0.00 -0.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   325  0.00 -0.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   331  0.00 -0.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   337  0.00 -0.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   343  0.00 -0.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   349  0.00 -0.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   355  0.00 -0.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   360  0.00 -0.00  0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False

Process finished with exit code 0
```

**Comment by pvcraven** [comment]
Here I hit the right analog trigger, which is sometimes used in steering. It fires, but only to the left.

The shoot 180 output is a print that I added that shows the shooting angle.

```
C:\Users\craven\Desktop\WebServer\arcade\venv\Scripts\python.exe C:/Users/craven/Desktop/WebServer/arcade/arcade/examples/dual_stick_shooter.py
========== <pyglet.input.base.Joystick object at 0x154A5850>
x       0
y       0
z       0
rx      0
ry      0
rz      0
hat_x   0
hat_y   0
buttons [False, False, False, False, False, False, False, False, False, False]
========== Extra joy
button_controls                = [Button(raw_name=Button 0), Button(raw_name=Button 1), Button(raw_name=Button 2), Button(raw_name=Button 3), Button(raw_name=Button 4), Button(raw_name=Button 5), Button(raw_name=Button 6), Button(raw_name=Button 7), Button(raw_name=Button 8), Button(raw_name=Button 9)] (list)
buttons                        = [False, False, False, False, False, False, False, False, False, False] (list)
device                         = DirectInputDevice(name=Controller (Gamepad F310)) (DirectInputDevice)
hat_x                          = 0 (int)
hat_x_control                  = AbsoluteAxis(name=hat, raw_name=Hat Switch) (AbsoluteAxis)
hat_y                          = 0 (int)
hat_y_control                  = AbsoluteAxis(name=hat, raw_name=Hat Switch) (AbsoluteAxis)
rx                             = 0 (int)
rx_control                     = AbsoluteAxis(name=rx, raw_name=X Rotation) (AbsoluteAxis)
ry                             = 0 (int)
ry_control                     = AbsoluteAxis(name=ry, raw_name=Y Rotation) (AbsoluteAxis)
rz                             = 0 (int)
rz_control                     = None (NoneType)
x                              = 0 (int)
x_control                      = AbsoluteAxis(name=x, raw_name=X Axis) (AbsoluteAxis)
y                              = 0 (int)
y_control                      = AbsoluteAxis(name=y, raw_name=Y Axis) (AbsoluteAxis)
z                              = 0 (int)
z_control                      = AbsoluteAxis(name=z, raw_name=Z Axis) (AbsoluteAxis)
========== Extra joy.device
_device                        = <pyglet.libs.win32.dinput.IDirectInputDevice8 object at 0x0D7378F0> (IDirectInputDevice8)
_subtype                       = 512 (int)
_type                          = 21 (int)
controls                       = [AbsoluteAxis(name=y, raw_name=Y Axis), AbsoluteAxis(name=x, raw_name=X Axis), AbsoluteAxis(name=ry, raw_name=Y Rotation), AbsoluteAxis(name=rx, raw_name=X Rotation), AbsoluteAxis(name=z, raw_name=Z Axis), Button(raw_name=Button 0), Button(raw_name=Button 1), Button(raw_name=Button 2), Button(raw_name=Button 3), Button(raw_name=Button 4), Button(raw_name=Button 5), Button(raw_name=Button 6), Button(raw_name=Button 7), Button(raw_name=Button 8), Button(raw_name=Button 9), AbsoluteAxis(name=hat, raw_name=Hat Switch)] (list)
display                        = None (NoneType)
is_open                        = False (bool)
manufacturer                   = None (NoneType)
name                           = Controller (Gamepad F310) (str)
========== pprint joy
<pyglet.input.base.Joystick object at 0x154A5850>
========== pprint joy.device
DirectInputDevice(name=Controller (Gamepad F310))
Using joystick controls: DirectInputDevice(name=Controller (Gamepad F310))
     3  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
     8  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    14  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    20  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    26  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    32  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    38  0.00  0.00  0.00  0.00  0.00  0.00  0  0 False False False False False False False False False False
    44  0.00 -0.00 -0.05  0.00 -0.00  0.00  0  0 False False False False False False False False False False
Shoot 180.0
Shoot 180.0
    50  0.00 -0.00 -0.52  0.00 -0.00  0.00  0  0 False False False False False False False False False False
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
    56  0.00 -0.00 -1.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
    62  0.00 -0.00 -1.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
    68  0.00 -0.00 -0.05  0.00 -0.00  0.00  0  0 False False False False False False False False False False
    74  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
    80  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
    86  0.00 -0.00 -1.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
    91  0.00 -0.00 -1.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
    97  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
Shoot 180.0
Shoot 180.0
   103  0.00 -0.00 -0.73  0.00 -0.00  0.00  0  0 False False False False False False False False False False
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
   109  0.00 -0.00 -1.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
Shoot 180.0
   115  0.00 -0.00 -0.86  0.00 -0.00  0.00  0  0 False False False False False False False False False False
Shoot 180.0
Shoot 180.0
Shoot 180.0
   121  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   127  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   133  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   139  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   145  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   151  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   157  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   163  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   169  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   175  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   181  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   187  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   193  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False
   199  0.00 -0.00 -0.00  0.00 -0.00  0.00  0  0 False False False False False False False False False False

Process finished with exit code 0
```

**Comment by SirGnip** [comment]
Thanks for the detailed logs.  
Try replacing line 184 with this:
`shoot_x, shoot_y, shoot_angle = get_joy_position(self.joy.rx, self.joy.ry)`

The issue here is that there is a crazy amount of variation in joysticks and how they expose buttons and axes to the underlying joystick API.  So, any kind of hardcoded approach to joystick axes will not work 100% of the time.   For instance, `joy.rx` and `joy.ry` means "right analog stick" for your joystick.  But, for the three styles of joystick I tested this with, the right analog stick is `joy.z` and `joy.rz`.  Why?  Manufacturer differences.
So, short of creating a database mapping every joystick known to man, we are in the world where the user must configure "what is right-stick?  left-stick?".  This could be driven by a local config file.  But, too fiddly for new users.  I believe that leaves us with in-game calibration "press your movement stick to the left... right... up... down..." that are used to detect what axis is which (I've used this in other games and have used it myself).
But, I'm not sure how to make this "easy" for learners, as they are the main audience of arcade. Thoughs?
In the meantime, I'm going to see if I can work up an ultra-minimal in-game joystick calibration library and hook it into the dual-stick shooter example.
